mshv_vtl/tdx: Add assembly function for TDG.VP.ENTER

ramesh-thomas opened the pull request:
Replace inline assembly implementation with assembly function. This makes the stack handling and the RBP save/restore cleaner. This is also consistent with current upstream TDX implementations. The C function that calls this function does not require the "noinline" directive with this change.

ramesh-thomas committed:
mshv_vtl/tdx: Add assembly function for TDG.VP.ENTER

Replace inline assembly implementation with assembly function. This
makes the stack handling and the RBP save/restore cleaner. This is also
consistent with current upstream TDX implementations. The C function
that calls this function does not require the "noinline" directive with
this change.

Signed-off-by: Ramesh Thomas <ramesh.thomas@intel.com>

yamahata commented on drivers/hv/mshv_tdx.S:
```diff
@@ -0,0 +1,79 @@
+        pushq   %r15
+
+	pushq	%rdi
+
```
We need to clear r8-r15.

ramesh-thomas replied:
ok

yamahata commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -829,35 +829,22 @@ void mshv_tdx_request_cache_flush(bool wbnoinvd)
-	register u64 r13 asm("r13");
-	register u64 r14 asm("r14");
-	register u64 r15 asm("r15");
+	struct tdx_module_args args;
```
I'm not sure which is better to use struct tdx_module_args or tdx_tdg_vp_enter_exit_info.
which is consistent?

ramesh-thomas replied:
We don't want to use tdx_tdg_vp_enter_exit_info to pass rcx and rdx arguments so created tdx_module_args. We ca pass tdx_tdg_vp_enter_exit_info  as argument to receive the exit info values directly.  

The input parameters rcx and rdx, they can be either passed directly as 2 parameters or inside tdx_module_args as following. Any preference?
__tdg_vp_enter(u64 rcx, u64 rdx, tdx_tdg_vp_enter_exit_info  *tdx_exit_info);

or
__tdg_vp_enter(tdx_module_args *args, tdx_tdg_vp_enter_exit_info  *tdx_exit_info);

yamahata replied:
Personally __tdg_vp_enter(u64 rcx, u64 rdx, tdx_tdg_vp_enter_exit_info *tdx_exit_info); seems good because it doesn't require tedious move of values between two structures for the code will be smaller.

romank-msft commented on drivers/hv/mshv_tdx.S:
```diff
@@ -0,0 +1,79 @@
+ *
+ * Return status of TDCALL via RAX.
+ */
+SYM_FUNC_START(__tdg_vp_enter)
```
Does the tdcall clobber the direction flag? The x86_64 ABI requires it to be cleared on function entry and return (3.2.1)

ramesh-thomas replied:
Don't know whether it clobbers or saves/restores. I can save the flags, clear the DF before the call and then restore the flags to meet the ABI requirement. Note that I wouldn't clear the DF after the call exits and instead save/restore the flags to avoid interfering with existing gcc/linux kernel assumptions of the DF state

romank-msft replied:
Thanks!

yamahata replied:
tdcall preserves rflags. Please refer to Intel CPU Architectural Extensions Specification for instruction reference.

ramesh-thomas replied:
The x86 ABI requires DF to be cleared before a function is called and exited. The flag saving and restoring is to set the DF to what it was before the call to tdg_vp_enter function.

romank-msft commented on drivers/hv/mshv_tdx.S:
```diff
@@ -0,0 +1,79 @@
+ */
+SYM_FUNC_START(__tdg_vp_enter)
+
+        push %_ASM_BP
```
Could you explain why not `pushq`? Or this is a macro that figures out between `ebp` and `rbp` :)

ramesh-thomas replied:
Yes, the macro figures it out based on the architecture.

romank-msft commented on drivers/hv/mshv_tdx.S:
```diff
@@ -0,0 +1,79 @@
+        xorl %r13d, %r13d
+        xorl %r14d, %r14d
+        xorl %r15d, %r15d
+        xorl %ebx,  %ebx
```
1. I guess these `xorl` are meant not to leak the state (the zero extension will take case of the upper 32 bits)? Please add a comment as to why these `xorl`'s are needed.
2. If that's for scrubbing the state, `xorl %r12d, %r12d` is not needed as the later `popq %r12` overwrites `r12`. Same for few others.

romank-msft replied:
*fixed reg names in the comment

ramesh-thomas replied:
ok, I will add the comment and would remove clearing ones that get restored.

romank-msft replied:
Thanks!

yamahata replied:
> 1. I guess these `xorl` are meant not to leak the state (the zero extension will take case of the upper 32 bits)? Please add a comment as to why these `xorl`'s are needed.
> 2. If that's for scrubbing the state, `xorl %r12d, %r12d` is not needed as the later `popq %r12` overwrites `r12`. Same for few others.

The clearing must be before tdcall to avoid information leak.. Not after. It doesn't make sense clearing those registers after tdcall.

ramesh-thomas replied:
I will add the clearing code as mentioned earlier before tdcall. 

Do we need to also scrub the values in those registers after exiting the tdcall or not?

yamahata replied:
> I will add the clearing code as mentioned earlier before tdcall.
> 
> Do we need to also scrub the values in those registers after exiting the tdcall or not?

No. Why do we need to clear them after tdcall?

ramesh-thomas replied:
Don't we need to clear any possible secrets before transitioning to other untrusted components, say if it returns to L0?

yamahata replied:
Ok, it makes sense as a safe-guard. Please add such comment.

ramesh-thomas commented:
Updated with feedback from @yamahata and @romank-msft

romank-msft commented on drivers/hv/mshv_tdx.S:
```diff
@@ -0,0 +1,95 @@
+        xor	%r15, %r15
+
+	pushf
+	cld
```
If tdcall preserves the flags as it was claimed in another thread, this is not needed I think. No harm if left as-is thou :)

ramesh-thomas replied:
TDCALL preserves the flag as it was when it got called. This code saves and restores it because we are clearing DF. I am not sure if  gcc in the Linux kernel follows the x86 ABI of clearing DF. It might be selectively doing it if it finds REP / string instructions. I didn't want to interfere with its optimization logic. 

I do see at several places assembly code in the kernel is doing it the way it is done here when external components are getting called.

romank-msft replied:
Gotcha, thsnks for looking!

ramesh-thomas replied:
Just noticed that the compiler gives below warning about the cld being redundant. Maybe the compiler knows that the DF was never changed before this.  I am removing it to avoid the unnecessary build warning.  @romank-msft @yamahata let me know if you still want me to add it regardless of the warning.

"drivers/hv/mshv_tdx.o: warning: objtool: __tdg_vp_enter+0x3b: redundant CLD"

yamahata reviewed (commented):
Why do we need to move out this structure to x86 public part?
We can simply use offset rule.
Something like
```
$(obj)/aaa.h: $(obj)/aaa.s FORCE
    $(call filechk,offsets,__aaa_ASM_OFFSETS_H__)
```

yamahata commented:
Other than moving out the struct, it looks good to me overall.

ramesh-thomas commented:
Thanks @yamahata and @romank-msft for the feedback. I have updated the commit with them.

yamahata commented:
> Thanks @yamahata and @romank-msft for the feedback. I have updated the commit with them.

Thanks for fixing offsets part. Now it looks good to me.

allenpais commented:
@chris-oo Could you please review the changes too. Thanks.

chris-oo reviewed (approved):
LGTM. Thanks for making this change.

ramesh-thomas commented:
@allenpais updated the commit with fix for the arm build error https://github.com/microsoft/OHCL-Linux-Kernel/actions/runs/13646277549/job/38275003068

benhillis commented:
@ramesh-thomas - sorry, looks like when I cleaned up the dev branch this needs to be rebased. Could you please do that?

ramesh-thomas commented:
> @ramesh-thomas - sorry, looks like when I cleaned up the dev branch this needs to be rebased. Could you please do that?

I have rebased to the latest project/hcl-dev/6.6 branch. Can you try?

Note that even without my change, the branch now shows a different behavior than before you did the rebase.  The guest VM shuts down after some time and the shell closes with the following message -
"Linux HCL 6.6.63-microsoft-hcl+ #1 SMP PREEMPT_DYNAMIC Fri Mar 28 12:22:34 PDT 2025 x86_64 x86_64 x86_64 GNU/Linux
sh-5.1# lsError: channel closed"

romank-msft commented:
> > @ramesh-thomas - sorry, looks like when I cleaned up the dev branch this needs to be rebased. Could you please do that?
> 
> I have rebased to the latest project/hcl-dev/6.6 branch. Can you try?
> 
> Note that even without my change, the branch now shows a different behavior than before you did the rebase. The guest VM shuts down after some time and the shell closes with the following message - "Linux HCL 6.6.63-microsoft-hcl+ #1 SMP PREEMPT_DYNAMIC Fri Mar 28 12:22:34 PDT 2025 x86_64 x86_64 x86_64 GNU/Linux sh-5.1# lsError: channel closed"

restarted the CI
